process_report/process_report.py

```diff
@@ -301,22 +253,40 @@ def main():
             bucket = get_invoice_bucket()
             invoice.export_s3(bucket)
 
-    billable_projects = remove_non_billables(merged_dataframe, pi, projects)
-    billable_projects = validate_pi_names(billable_projects)
-
     if args.upload_to_s3:
         backup_to_s3_old_pi_file(old_pi_file)
-    credited_projects = apply_credits_new_pi(billable_projects, old_pi_file)
 
-    export_billables(credited_projects, args.output_file)
-    export_pi_billables(credited_projects, args.output_folder, invoice_month)
-    export_BU_only(billable_projects, args.BU_invoice_file, args.BU_subsidy_amount)
-    export_HU_BU(credited_projects, args.HU_BU_invoice_file)
+    billable_inv = billable_invoice.BillableInvoice(
+        name=args.output_file,
+        invoice_month=invoice_month,
+        data=merged_dataframe.copy(),
+        nonbillable_pis=pi,
+        nonbillable_projects=projects,
+        old_pi_filepath=old_pi_file,
+    )
+    billable_inv.process()
+    billable_inv.export()
+    if args.upload_to_s3:
+        bucket = get_invoice_bucket()
+        billable_inv.export_s3(bucket)
+
+    bu_internal_inv = bu_internal_invoice.BUInternalInvoice(
+        name=args.BU_invoice_file,
+        invoice_month=invoice_month,
+        data=billable_inv.data,
```

Comment:
will this modify the `billable_inv` dataframe? The reason I ask this is because we could consolidate all the s3 exports later. something like:

```python
if args.upload_to_s3:
    bucket = get_invoice_bucket()
    billable_inv.export_s3(bucket)
    bu_internal_inv.export_s3(bucket)
 ```

Reply:
also, it's just good to understand if this does modify it because then it would mean that after this runs "billable_inv" is no longer what it says (it may have turned into bu_internal_inv)

Reply:
If I don't explicitly call `copy()` on a dataframe, then you can assume that the same dataframe is being assigned by reference:
```
a = pd.DataFrame({'a': [1,2,3,4],
                  'b': [1,2,3,4]})
b = a
b.loc[0, 'a'] = 100
print(a)
print(b)>>> 
>>> a = pd.DataFrame({'a': [1,2,3,4],
...                   'b': [1,2,3,4]})
>>> b = a
>>> b.loc[0, 'a'] = 100
>>> print(a)
     a  b
0  100  1
1    2  2
2    3  3
3    4  4
>>> print(b)
     a  b
0  100  1
1    2  2
2    3  3
3    4  4 
```

I will make sure each invoice's dataframes won't reference each other when exported together like in your suggestion.

Reply:
@naved001  I'll consolidate all the `export()` and `export_s3()` calls together after your detailed review.

Reply:
@QuanMPhm thanks, this largely looks fine to me.

Reply:
While consolidating the export statements, I thought that it would make the code more organized if I process and export the invoices in the same code block. So instead of consolidating all the `export()` calls together, I defined [`process_and_export_invoices`](https://github.com/CCI-MOC/process_csv_report/pull/71/files#diff-fd9ba5f8a61085103fd99900c2be8be11438be8fb9919be0868c50b4a5df59b0R65-R71) to process and export a list of invoices. 

For now, `process_and_export_invoices` is called twice. First to export the nonbillable, billable, and lenovo invoice. Second to export the HU-BU and BU-Internal, since these two depends on processing that's done in the billable invoice.